enhancements/authentication/configuring-webhook-token-authenticators.md

```diff
@@ -0,0 +1,169 @@
+
+1. Make it possible to configure an endpoint capable of performing kube token reviews
+2. Modify cluster-authentication-operator so that it configures authentication via the OpenShift OAuth
+   server by using the `webhookTokenAuthneticators` field instead of the current patching of the
+   kube-apiserver.
+
+### Non-Goals
+
+1. provide a partial authentication operator to any external authenticator.  Once the standard oauth-server
+   is disabled, authentication configuration is *entirely* the responsibility of the new authentication stack.
+2. allow both the built-in oauth-server and another authentication stack to be enabled at the same time.
+
+## Proposal
+
+### Implementation
+
+#### API
+
+`authentications.config.openshift.io/v1` already contains the `spec.webhookTokenAuthenticators`. It
+expects an array of references to secrets in the `openshift-config` namespace. The field is currently
+unused so it is safe to leverage it for the purposes of configuring webhook token authenticators.
+
+Any secret reference which gets set to this field has to be validated. Such a referenced secret will be
+expected to contain a `kubeconfig` key with a valid kubeconfig configuration for the webhook. In 4.4,
+it will only be possible to configure a single element for the `wehbhookTokenAuthenticators` field
+and the referenced kubeconfigs will only be allowed to use the `-data` forms of their fields to configure
+certificate authority, client certificate and key.
+
+This path is compatible with kubernetes as it is possible to directly map the referenced
+secrets' contents to files and specify these in the kube-apiserver configuration.
+
+#### Backend - kube-apiserver-operator
+
+1. observes any changes in `webhookTokenAuthenticators` field of the `authentication/cluster` resource
+  and performs validation of any references in that field
+2. it pushes any webhook authenticator that passes the validation to a secret in `openshift-kube-apiserver`
+  namespace, the name of the secret is derived from its index in the `spec.webhookTokenAuthenticators`
+  field of the `authentications.config/cluster` object such that `name=webhook-authenticator-<index>`
+3. if no webhook authenticators pass validation, the operator keeps the previous state
+4. redeploys the kube-apiserver static pods while specifying the new
+  mounted files in the `webhookTokenAuthenticators` [config field](https://github.com/openshift/api/blob/992bc3a41fe6c4e5bcf27dc8080f9e5c114c33fc/kubecontrolplane/v1/types.go#L146).
+
+#### Backend - Integrated OAuth Server Token Authentication
+
+1. cluster-authentication-operator observes `authentication.config/cluster` `type` field
+2. if set to "IntegratedOAuth", it creates an `openshift-config/webhook-authentication-integrated-oauth`
+  secret and adds it to the `webhookTokenAuthenticators` field of the `authentication.config/cluster`
+  resource
+  - the above secret contains details that point kube-apiserver to the oauth-apiserver for token reviews
```

Comment:
so it also watches the config and sets it back to `openshift-config/webhook-authentication-integrated-oauth` if changed by the admin?

Reply:
correct